### PR TITLE
Upgrade zio-kafka + apply RCA to tests

### PR DESCRIPTION
**Pull request by gurghet**
This PR upgrade zio-kafka and update tests to make them characterize better real-world edge cases.

### CODE REVIEW

File: core/src/test/scala/tamer/utils/FakeConsumer.scala [hunk 1]
```diff
@@ -108,18 +108,34 @@ sealed class FakeConsumer[IK, IV](
...
+                (
+                  new TopicPartition("topic", committableRecord.partition),
+                  new ProducerRecord("topic", committableRecord.key.asInstanceOf[IK], committableRecord.value.asInstanceOf[IV])
+                )
+              )
+            ) *>
+              ZStream(committableRecord, committableRecord.copy())
+          } else
+            ZStream(committableRecord)
+        }
```
Comment (gurghet) [inline core/src/test/scala/tamer/utils/FakeConsumer.scala:138]: This would open to testing mitigations to current async bugs through unit testing, which is appealing. Corresponding functions however would have to be written as well: for example if we used `endOffsets` to decide whether to commit or not, we would have to update it because only works when starting up right now, it doesn't keep track of the fictitious offset. So right now it's only useful against code regressions.

File: core/src/test/scala/tamer/utils/FakeConsumer.scala [hunk 2]
```diff
@@ -108,18 +108,34 @@ sealed class FakeConsumer[IK, IV](
             record = new ConsumerRecord("topic", 0, offset, record.key().asInstanceOf[K], record.value().asInstanceOf[V]),
             commitHandle = (toCommit: Map[TopicPartition, Long]) =>
               (for {
-                r <- Task(new java.util.Random().nextInt(1)).map(_.longValue) // simulates the asynchronicity
-                _ <- Task(Thread.sleep(r)) // simulates the asynchronicity
```
Comment (gurghet) [inline core/src/test/scala/tamer/utils/FakeConsumer.scala:112]: While it's true that this action is asynchronous, the effects often where not measurable.

File: core/src/main/scala/tamer/Tamer.scala [hunk 3]
```diff
@@ -341,10 +355,10 @@ object Tamer {
         .make(consumerSettings)
         .mapError(TamerError("Could not make state consumer", _))
       val stateProducer = Producer
-        .make(producerSettings, stateKeySerde.serializer, serdes.stateSerde)
+        .make(producerSettings)
         .mapError(TamerError("Could not make state producer", _))
       val valueProducer = Producer
-        .make(producerSettings, serdes.keySerializer, serdes.valueSerializer)
+        .make(producerSettings)
         .mapError(TamerError("Could not make value producer", _))
```
Comment (sirocchj) [inline core/src/main/scala/tamer/Tamer.scala:362]: I wonder if we should just now use one producer

File: core/src/main/scala/tamer/Tamer.scala [hunk 4]
```diff
@@ -289,8 +298,13 @@ object Tamer {
...
     ) = for {
-      fiber <- sinkStream(kvStream, sinkRegistry, log)
-      _     <- sourceStream(kvChunkQueue, stateRegistry, log).ensuringFirst(stopSource(log)).ensuring(drainSink(fiber, kvStream, sinkRegistry, log))
+      fiber <- sinkStream(kvStream, serdes.keySerializer, serdes.valueSerializer, log).provideSomeLayer[Clock](sinkRegistry)
+      _ <- sourceStream(kvChunkQueue, stateRegistry, log)
+        .ensuringFirst(stopSource(log))
+        .ensuring(
+          drainSink(fiber, kvStream, serdes.keySerializer, serdes.valueSerializer, log)
+            .provideSomeLayer[Clock](sinkRegistry)
+        )
```
Comment (sirocchj) [inline core/src/main/scala/tamer/Tamer.scala:307]: I like providing the layer here (it was done in a similar fashion earlier), or ever further down!, but at the moment we have two layers for the same type. One we're now providing here, the other we're passing like a regular parameter (I know, I know!) and we're finally providing it in the `sourceStream` method.

I'm starting to wonder if we shouldn't find a way to tag these types so they are different to the compiler and we just provide them at the edge. Wdyt?

File: core/src/test/scala/tamer/utils/FakeConsumer.scala [hunk 5]
```diff
@@ -22,7 +21,8 @@ sealed class FakeConsumer[IK, IV](
     val inFlight: Queue[(TopicPartition, ProducerRecord[IK, IV])],
     val produced: Map[TopicPartition, Queue[ProducerRecord[IK, IV]]],
     log: LogWriter[Task]
-) extends Service {
+) extends Consumer {
+  val r = new java.util.Random()
```
Comment (sirocchj) [inline core/src/test/scala/tamer/utils/FakeConsumer.scala:25]: layer?

File: core/src/test/scala/tamer/utils/FakeConsumer.scala [hunk 6]
```diff
@@ -7,13 +7,12 @@ import org.apache.kafka.common.{Metric, MetricName, PartitionInfo, TopicPartitio
 import tamer.utils.FakeConsumerUtils._
 import zio.clock.Clock
 import zio.duration.Duration
-import zio.kafka.consumer.Consumer.Service
-import zio.kafka.consumer.{CommittableRecord, SubscribedConsumer, Subscription}
+import zio.kafka.consumer.{CommittableRecord, Consumer, SubscribedConsumer, Subscription}
 import zio.kafka.serde.Deserializer
 import zio.random.Random
 import zio.stream.ZStream
```
Comment (sirocchj) [inline core/src/test/scala/tamer/utils/FakeConsumer.scala:13]:
```suggestion
import zio.stream.{Stream, ZStream}
```

File: core/src/test/scala/tamer/utils/FakeConsumer.scala [hunk 7]
```diff
@@ -139,6 +155,10 @@ sealed class FakeConsumer[IK, IV](
   override def subscribeAnd(subscription: Subscription): SubscribedConsumer                                                             = ???
   override def subscription: Task[Set[String]]                                                                                          = ???
   override def metrics: Task[Map[MetricName, Metric]]                                                                                   = ???
+  override def partitionedAssignmentStream[R, K, V](
+      keyDeserializer: Deserializer[R, K],
+      valueDeserializer: Deserializer[R, V]
+  ): stream.Stream[Throwable, Chunk[(TopicPartition, ZStream[R, Throwable, CommittableRecord[K, V]])]] = ???
```
Comment (sirocchj) [inline core/src/test/scala/tamer/utils/FakeConsumer.scala:161]:
```suggestion
  ): Stream[Throwable, Chunk[(TopicPartition, ZStream[R, Throwable, CommittableRecord[K, V]])]] = ???
```
to fix also in `partitionedStream` method signature

File: core/src/test/scala/tamer/utils/FakeConsumer.scala [hunk 8]
```diff
@@ -7,13 +7,12 @@ import org.apache.kafka.common.{Metric, MetricName, PartitionInfo, TopicPartitio
...
 import zio.duration.Duration
-import zio.kafka.consumer.Consumer.Service
-import zio.kafka.consumer.{CommittableRecord, SubscribedConsumer, Subscription}
+import zio.kafka.consumer.{CommittableRecord, Consumer, SubscribedConsumer, Subscription}
 import zio.kafka.serde.Deserializer
 import zio.random.Random
 import zio.stream.ZStream
 import zio.stream.ZStream.repeatEffectOption
-import zio.{Cause, IO, Queue, Ref, Schedule, Task, UIO, URIO, ZIO, ZQueue, stream}
+import zio.{Cause, Chunk, IO, Queue, Ref, Schedule, Task, UIO, URIO, ZIO, ZQueue, stream}
```
Comment (sirocchj) [inline core/src/test/scala/tamer/utils/FakeConsumer.scala:15]:
```suggestion
import zio.{Cause, Chunk, IO, Queue, Ref, Schedule, Task, UIO, URIO, ZIO, ZQueue}
```

**Review by sirocchj — DISMISSED** [review]
Awesome!! Some minor nit and a couple of suggestions for further polishing

File: core/src/main/scala/tamer/Tamer.scala (hunk 3 above)
Comment (gurghet) [reply core/src/main/scala/tamer/Tamer.scala:362]: Yeah I think it fits with the purpose of the PR, I'll do it

Comment (sirocchj) [reply core/src/main/scala/tamer/Tamer.scala:362]: Great!! Probably also worth calling them just `consumer` and `producer` at this stage :)

File: core/src/test/scala/tamer/utils/FakeConsumer.scala (hunk 5 above)
Comment (gurghet) [reply core/src/test/scala/tamer/utils/FakeConsumer.scala:25]: I think this was because I could not override and constrain `R` at the same time but I'll have a second look

File: core/src/main/scala/tamer/Tamer.scala (hunk 4 above)
Comment (gurghet) [reply core/src/main/scala/tamer/Tamer.scala:307]: Why do they need to be tagged? I don't understand. Is it because we allow using different registries for different topics?

Comment (gurghet) [reply core/src/main/scala/tamer/Tamer.scala:307]: Ok, re-reading the code it makes sense. I just didn't realized it worked like that.

Comment (gurghet) [reply core/src/main/scala/tamer/Tamer.scala:306]: Among the many options I just used type aliases for this. I initially tried with shapeless but it made the code 5x more complicated, case classes also seemed bulky, I thought about zio subtypes but didn't want to add a new dependency.

Comment (sirocchj) [reply core/src/main/scala/tamer/Tamer.scala:306]: In hindsight I believe we only need `Has[Registry]` and topic name should just be supplied as a normal parameter to the `serializer` method. Also, `deserializer` should have never required anything more than `Has[Registry]`.
I am happy to do this cleanup myself after this PR is merged.

Comment (gurghet) [reply core/src/main/scala/tamer/Tamer.scala:306]: > Also, deserializer should have never required anything more than Has[Registry].
True just keep in mind they need to have the same `R` when passed to create a serde.

**Review by barambani — APPROVED** [review]

**Review by sirocchj — APPROVED** [review]
